lang/cs/Org.Apache.REEF.IO/FileSystem/Hadoop/HDFSCommandRunner.cs

```diff
@@ -0,0 +1,172 @@
+                var hadoopHomeFromEnv = Environment.GetEnvironmentVariable("HADOOP_HOME");
+                if (null == hadoopHomeFromEnv)
+                {
+                    throw new Exception("HADOOP_HOME not set and no path to the hadoop installation provided.");
+                }
+                _hdfsCommandPath = Path.Combine(hadoopHomeFromEnv, "bin", "hdfs.cmd");
+            }
+            else
+            {
+                _hdfsCommandPath = Path.Combine(hadoopHome, "bin", "hdfs.cmd");
+            }
+
+
+            if (!File.Exists(_hdfsCommandPath))
+            {
+                throw new Exception("HDFS command does not exist: " + _hdfsCommandPath);
+            }
+        }
+
+        internal CommandResult Run(string hdfsCommandLineArguments)
+        {
+            var outList = new List<string>();
+            var errList = new List<string>();
+            var tries = _numberOfRetries;
+            var origTries = _numberOfRetries;
+            Process process;
+            do
+            {
+                outList.Clear();
+                Logger.Log(Level.Info, "Trial {0} Timeout in {1} secs Executing: {2}", origTries - tries,
+                    _timeOutInMilliSeconds/1000,
+                    _hdfsCommandPath + " " + hdfsCommandLineArguments);
+                var startInfo = new ProcessStartInfo
+                {
+                    FileName = _hdfsCommandPath,
+                    Arguments = hdfsCommandLineArguments,
+                    UseShellExecute = false,
+                    RedirectStandardOutput = true,
+                    RedirectStandardError = true
+                };
+
+                process = Process.Start(startInfo);
+
+                if (process == null)
+                {
+                    throw new Exception(string.Format("HDFS cmd {0} {1} process didn't start.", _hdfsCommandPath,
+                        hdfsCommandLineArguments));
+                }
+
+                process.OutputDataReceived += delegate(object sender, DataReceivedEventArgs e)
+                {
+                    if (!string.IsNullOrWhiteSpace(e.Data))
+                    {
+                        outList.Add(e.Data.Trim());
+                    }
+                };
+
+                process.ErrorDataReceived += delegate(object sender, DataReceivedEventArgs e)
+                {
+                    if (!string.IsNullOrWhiteSpace(e.Data))
+                    {
+                        errList.Add(e.Data.Trim());
+                    }
+                };
+                process.BeginErrorReadLine();
+                process.BeginOutputReadLine();
+
+                if (process.WaitForExit(_timeOutInMilliSeconds))
+                {
+                    break;
+                }
+
+                Logger.Log(Level.Warning, "Process has not exited after a timeout of {0} secs. So killing it",
+                    _timeOutInMilliSeconds);
+                process.Kill();
+            } while (--tries > 0);
```

Comment:
Uhm, because this made more sense to @shravanmn when he wrote the code when he wrote the code I ~~blatantly ripped off~~ used as generous inspiration for this? :-)

Seriously, though: I agree with the sentiment that a `for` loop would be easier to understand here and will change the code to use one.